packages/react-router/modules/Match.js

```diff
@@ -0,0 +1,98 @@
+import React from "react";
+import PropTypes from "prop-types";
+import warning from "tiny-warning";
+import RouterContext from "./RouterContext";
+import matchPath from "./matchPath";
+
+/**
+ * The public API for matching a route and providing that match to children.
+ */
+class Match extends React.Component {
+  render() {
+    return (
+      <RouterContext.Consumer>
+        {context => {
+          if (!context) {
+            throw new Error(
+              __DEV__
+                ? "You should not use <Match> outside a <Router>"
+                : "Invariant failed"
+            );
+          }
+          const location = this.props.location || context.location;
+          let match = this.props.computedMatch
+            ? this.props.computedMatch // <Switch> already computed the match for us
+            : this.props.path
+              ? matchPath(location.pathname, this.props)
+              : context.match;
+
+          const props = { ...context, location, match };
+
+          let { children } = this.props;
+
+          // Preact uses an empty array as children by
+          // default, so use null if that's the case.
+          if (Array.isArray(children) && children.length === 0) {
+            children = null;
+          }
+
+          if (typeof children === "function") {
+            children = children(props);
+
+            if (children === undefined) {
+              if (__DEV__) {
+                const { path } = this.props;
+
+                warning(
+                  false,
+                  "You returned `undefined` from the `children` function of " +
+                    `<Match${path ? ` path="${path}"` : ""}>, but you ` +
+                    "should have returned a React element or `null`"
+                );
+              }
+
+              children = null;
+            }
+          }
+
+          if (children === null) {
+            return null;
+          }
+          return (
+            <RouterContext.Provider value={props}>
+              {children}
+            </RouterContext.Provider>
+          );
+        }}
+      </RouterContext.Consumer>
+    );
+  }
+}
+
+if (__DEV__) {
+  Match.propTypes = {
+    children: PropTypes.oneOfType([PropTypes.func, PropTypes.node]),
+    exact: PropTypes.bool,
+    location: PropTypes.object,
+    path: PropTypes.oneOfType([
+      PropTypes.string,
+      PropTypes.arrayOf(PropTypes.string)
+    ]),
+    sensitive: PropTypes.bool,
+    strict: PropTypes.bool
+  };
+
+  Match.prototype.componentDidUpdate = function(prevProps) {
```

Comment:
We could de-duplicate this (accross Route, Switch and Match) with the following. But constructing that message from the ```prototype.constructor.name``` feels a bit icky.

```js
export function addLocationPropWarning(prototype){
	prototype.componentDidUpdate = function(prevProps) {
    warning(
      !(this.props.location && !prevProps.location),
      `<${prototype.constructor.name}> elements should not change from uncontrolled to controlled (or vice versa). You initially used no "location" prop and then provided one on a subsequent render.`
    );

    warning(
      !(!this.props.location && prevProps.location),
      `<${prototype.constructor.name}> elements should not change from controlled to uncontrolled (or vice versa). You provided a "location" prop initially but omitted it on a subsequent render.`
    );
  };
}

// Match.js:
addLocationPropWarning(Match.prototype);
```

Reply:
If you don't like the idea with the name retrieving by the constructor name, just add a second parameter for the name and use this as replacement for `prototype.constructor.name`.

Reply:
Well, that would be entirely to obvious... I think I need more sleep. Thanks for the pointer though.

Reply:
Anyway, I think the function would be better named `addControlledUncontrolledChangeWarning`, even so the change be vice versa. In this case it is possible to use this function also for other properties. Currently this is not the case but my crystal bead is currently in repair and so I cannot see what the future will bring.